### PR TITLE
ActiveRecord: Optimize cache_key computation

### PR DESCRIPTION
**Pull request by skryukov**
While a request optimization, we added a cache, but it wasn't as fast as we hoped. It turned out that queries with `distinct_value` are executed while cache key computation. This PR addresses this performance issue.

### CODE REVIEW

File: activerecord/lib/active_record/relation.rb [hunk 1]
```diff
@@ -369,7 +369,9 @@ def compute_cache_version(timestamp_column) # :nodoc:
         select_values = "COUNT(*) AS #{connection.quote_column_name("size")}, MAX(%s) AS timestamp"
 
         if collection.has_limit_or_offset?
-          query = collection.select("#{column} AS collection_cache_key_timestamp")
+          query = collection.spawn
+          query = query.select(table[Arel.star]) if distinct_value && query.select_values.empty?
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/relation.rb:373]:
```suggestion
          query = query.select!(table[Arel.star]) if distinct_value && query.select_values.empty?
```

File: activerecord/lib/active_record/relation.rb [hunk 2]
```diff
@@ -369,7 +369,9 @@ def compute_cache_version(timestamp_column) # :nodoc:
...
+          query = query.select("#{column} AS collection_cache_key_timestamp")
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/relation.rb:374]:
```suggestion
          query = query.select!("#{column} AS collection_cache_key_timestamp")
```

**Review by rafaelfranca — COMMENTED** [review]
Can you add a test for this optimization? We could check the SQL generated.

**Comment by skryukov** [comment]
@rafaelfranca Thanks for the review! I've added an assertion to check if the relation was loaded.

**Comment by rafaelfranca** [comment]
Thank you! Can you open a new PR for the other commit?

**Pull request merged**